### PR TITLE
doc(Widgets): Add Widgets development documentation

### LINKED ISSUE #682: Widgets refactoring
Following-up from discussion notes here:
https://docs.google.com/document/d/1lbFj4N68PhuFC8NElz_7wVW655mMdxRdP8Qkn17wwA4

cc: @lassoan  @floryst @jcfr @martinken @thewtex @aylward @timthirion

## Background
Widgets in vtk-js are at a very early stage. We brought up some issues with the current architecture, and believe now is the time to address some of them since the number of widgets in vtk.js is still very limited.

## Identified Issues
* Confusions between roles of widget and widgetRepresentation leads to unnecessary complexities
* Lack of separation between representation and source limits data sharing between multiple representations and forces duplications and syncing of state
* Very costly to have a lot of handle widgets

## Proposed solutions
- [ ] Separate widget architecture into 3 classes for:
  - interaction (listening currently done in widget, and handling currently done in widgetRep),
  - representation (currently actor and mapper in widgetRep)
  - state/data (currently source in widgetRep).
- [ ] Base handlewidget on glyphs to optimize the creation of a high number of similar widgets (?)

*Please edit with suggestions...*

### PR DESCRIPTION
**Pull request by finetjul**
Issue #1225

Fixes #682

### CODE REVIEW

**Comment by floryst** [comment]
Nice! I'll read it in more detail in a bit. Right now, I think there are some overlaps with the current widget docs https://kitware.github.io/vtk-js/docs/concepts_widgets.html, so I think it's worthwhile to merge them into a larger document since yours is more detailed in terms of actual dev docs.

File: Documentation/content/docs/develop_widget.md [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+
+A widget can be added to multiple widget managers. Each widget manager will return a `handle` for its specific renderer.
+
+### Widget suppression
+
+Removing a widget from a view is done by `widgetManager.removeWidget(widget)`. Removing a widget from a widget manager also removes all its representations and the widget cannot receive events anymore. Once a widget is removed from all widget managers, it can be safely deleted by `widget.delete()`. Deleting a widget before removing it from every widget managers will cause some issues because these managers will still try to use it.
+
+### Widget state
+
+The widget state stores widget data used accross renderers. The widget state is composed of sub-states(i.e. Object) made of properties that describe particular aspects of the widget. For example, sub-states can store positions and sizes of actors to be rendered.
```
Comment (mix3d) [inline Documentation/content/docs/develop_widget.md]: typo: across

File: Documentation/content/docs/develop_widget.md [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+
+Each widget manager can give the focus to at most one widget at any given time. The focus defines which widget is active and should handle events. Note that it is still possible to respond to events without having the focus (see Widget Behavior). Focus is given to the widget using `widgetManager.grabFocus(widget)`.
+A call to `widgetManager.enablePicking()` allows the user to interact with widgets. It releases the focus of the previously holding widget if any.
+
+### Widget creation
+
+A widget is created by `vtkWidget.newInstance(INITIAL_VALUES)` where `vtkWidget` is the base class of the widget being created and `INITIAL_VALUES` are the arguments given to the constructor of the widget. Note that some parameters cannot be changed later on. `INITIAL_VALUES` may be the only way to set those parameters. During this call the "widget state" is created. It is unique to a widget instance. This allows for synchronization of the same widget accross renderers (a change of the widget state in one view is directly rendered in all other views).
```
Comment (thewtex) [inline Documentation/content/docs/develop_widget.md]: accross -> across

File: Documentation/content/docs/develop_widget.md [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+
+### Widget behavior
+
+The widget behavior is the place where the logic of the widget happens. The widget behavior is the handle returned when a widget is added to a widget manager. The widget behavior receives and responds to mouse and keyboard events through handler methods. These methods are named `handle{XXX}(callData)` where `XXX` is the name of the event (like `KeyPress`, `KeyUp`, `MouseMove`, `LeftButtonPress`, etc...) and `callData` is the event data (it contains informations like the mouse position, the keyboard state ...). All events don't need a handler method: if no handler is provided the widget behavior ignores the event. Each handler must return either `macro.EVENT_ABORT` or `macro.VOID`. `macro.EVENT_ABORT` means that the event should not be propagated to other widgets wherehas `macro.VOID` means that the event should be propagated. Note that the order in which widgets receive events is not guaranteed, so returning the wrong value might starve other widgets from events they expect.
```
Comment (thewtex) [inline Documentation/content/docs/develop_widget.md]: wherehas -> whereas

**Comment by mix3d** [comment]
should this close #682 as well?

**Comment by floryst** [comment]
@mix3d Yeah it should. Thanks for catching that.

**Pull request merged**